sdk/provisioning/Azure.Provisioning/src/Primitives/ProvisioningConstruct.cs

```diff
@@ -0,0 +1,201 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using Azure.Provisioning.Expressions;
+
+namespace Azure.Provisioning.Primitives;
+
+/// <summary>
+/// A named Bicep entity, like a resource or parameter.
+/// </summary>
+public abstract class NamedProvisioningConstruct : ProvisioningConstruct
+{
+    /// <summary>
+    /// Gets the the Bicep name of the resource.  This can be used to refer to
+    /// the resource in expressions, but isn't the Azuree name of the resource.
+    /// </summary>
+    public string ResourceName { get; private set; }
```

Comment:
`NamedProvisioningConstruct` does one better and gives it a resource name.  This is the base for everything that'll be top-level in your Bicep file like `resource`s or `param`s.